### PR TITLE
GCI-142 Check bad code practices through Findbugs or Sonar and fix those issues. (2)

### PR DESCRIPTION
**Pull request by PermissionError**
## Description of what I changed

Check bad code practices through Findbugs or Sonar and fix those issues. (Exposing internal representations of mutableobjects)

## Issue I worked on

see https://issues.openmrs.org/browse/GCI-142

## Checklist: I completed these to help reviewers :)

- [x] My pull request only contains **ONE single commit**
(the number above, next to the 'Commits' tab is 1).

  No? -> [read here](https://wiki.openmrs.org/display/docs/Pull+Request+Tips) on how to squash multiple commits into one

- [x] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [ ] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [x] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [x] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [x] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

### CODE REVIEW

File: api/src/main/java/org/openmrs/BaseConceptMap.java [hunk 1]
```diff
@@ -79,31 +79,45 @@ public void setChangedBy(User changedBy) {
...
 	 * @param dateCreated the dateCreated to set
 	 */
 	@Override
 	public void setDateCreated(Date dateCreated) {
-		this.dateCreated = dateCreated;
+		if(dateCreated == null) {
+			this.dateCreated = null;
+			return;
+		}
+		this.dateCreated = (Date) dateCreated.clone();
```
Comment (PrathameshMutkure) [inline api/src/main/java/org/openmrs/BaseConceptMap.java]: Hii @PermissionError, have a look at [this](http://www.informit.com/articles/article.aspx?p=31551&seqNum=2) article, using `clone()` method in setters may possess some security risks :)

File: api/src/main/java/org/openmrs/BaseConceptMap.java [hunk 2]
```diff
@@ -79,31 +79,45 @@ public void setChangedBy(User changedBy) {
...
 	 * @param dateChanged the dateChanged to set
 	 */
 	@Override
 	public void setDateChanged(Date dateChanged) {
-		this.dateChanged = dateChanged;
+		if(dateChanged == null) {
+			this.dateChanged = null;
+			return;
+		}
+		this.dateChanged = (Date) dateChanged.clone();
```
Comment (PrathameshMutkure) [inline api/src/main/java/org/openmrs/BaseConceptMap.java]: Same as above

File: api/src/main/java/org/openmrs/BaseConceptMap.java (hunk 1 above)
Comment (PermissionError) [reply api/src/main/java/org/openmrs/BaseConceptMap.java]: Thanks for the tip, very valuable indeed. PR amended :)

**Comment by dkayiwa** [comment]
I suggest that each pull request deals with only one category of changes, instead of mixing them.

**Comment by PermissionError** [comment]
@dkayiwa Re-opened as #3096 & #3097

**Comment by dkayiwa** [comment]
@PermissionError cab you make the commit message reflect what exactly you are doing in the pull request? That is, instead of the generic `Check bad code practices through Findbugs or Sonar and fix those issues`

**Comment by PermissionError** [comment]
@dkayiwa Thanks for the feedback, should be more specific now.

**Pull request closed without merging**